tokio/src/sync/notify.rs

```diff
@@ -319,17 +348,51 @@ impl Notify {
     }
 
     /// Notifies all waiting tasks
-    pub(crate) fn notify_waiters(&self) {
+    ///
+    /// If a task is currently waiting, that task is notified. Unlike with
+    /// `notify()`, no permit is stored to be used by the next call to
+    /// [`notified().await`]. The purpose of this method is to notify all
+    /// already registered waiters. Registering for notification is done by
+    /// acquiring an instance of the `Notified` future via calling `notified()`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let notify = Arc::new(Notify::new());
+    ///     let notify2 = notify.clone();
+    ///
+    ///     let notified1 = notify.notified();
+    ///     let notified2 = notify.notified();
+    ///
+    ///     let handle = tokio::spawn(async move {
+    ///         println!("sending notifications");
+    ///         notify2.notify_waiters();
+    ///     });
+    ///
+    ///     notified1.await;
+    ///     notified2.await;
+    ///     println!("received notifications");
+    /// }
+    /// ```
+    pub fn notify_waiters(&self) {
```

Comment:
It looks like the current implementation calls `wake()` while holding the lock. This is problematic as it can result in a deadlock.

Consider a task that is backed by an [`ArcWake`](https://docs.rs/futures/0.3.8/futures/task/trait.ArcWake.html). The task is dropped only when all the wake handles go out of scope. Imagine if the drop implementation of a task calls `notify_waiters()` it self. The `notify_waiters()` function waits on the lock, however, the drop function got called by `wake()` in `notify_waiters()` itself (because the waker was the final ref to the Arc and that dropped). That results in an attempt to "re-enter" the mutex, which deadlocks.

The solution is to store the waker on the stack and notify it while not holding the mutex. See how [this](https://github.com/tokio-rs/tokio/blob/master/tokio/src/io/driver/scheduled_io.rs#L217) does it.